_posts/2025-04-23-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,202 @@
+
+image::mcp_server_oidc_devui.png[MCP Server and OIDC in DevUI,align="center"]
+
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
+
+What about other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] ?
+No problems, you can login to them from OIDC DevUI as well, for example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0] - the only requirement is to update your provider application registration to allow callbacks to DevUI.
+
+Now, after logging in with `Keycloak`, copy the acquired access token using a provided copy button:
+
+image::login_and_copy_access_token.png[Login and copy access token,align="center"]
+
+Now go the MCP Server card, select the `Tools` option and choose the `user-name-provider` tool:
+
+image::mcp_server_choose_tool.png[Choose MCP Server tool,align="center"]
+
+Paste the copied access token into the Tool's `Bearer token` field:
+
+image::mcp_server_bearer_token.png[MCP Server Bearer token,align="center"]
+
+Make a tool call by selecting the `Call` action and get the response:
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+=== Access MCP SSE server with curl
+
+Let's try to use `curl` as well and also learn a little bit how MCP SSE transport works. 
+
+Access the main SSE endpoint with the access token without the access token first:
+
+[source,shell]
+----
+curl -v localhost:8080/mcp/sse
+----
+
+You will get HTTP 401 error.
+
+Now, get and copy the access token as explained in the <<mcp-server-devui>> section above.
+
+[source,shell]
+----
+curl -v -H "Authorization: Bearer ey..." localhost:8080/mcp/sse
+----
+
+and get an SSE response such as:
+
+[source,shell]
+----
+< content-type: text/event-stream
+< 
+event: endpoint
+data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+The current `curl` SSE session must stay open, for it to receive updates related to dynamically allocated `data` endpoint.
+
+Open another window and use the same access token to post the tool call request to the `data` endpoint:
+
+[source,shell]
+----
+url -v -H "Authorization: Bearer ey..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+where a `call.json` looks like this:
+
+[source,json]
+----
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "method": "tools/call",
+  "params": {
+    "name": "user-name-provider",
+    "arguments": {
+    }
+  }
+}
+----
+
+Now look in the first curl window and observe a response such as:
+
+[source,shell]
+----
+event: message
+data: {"jsonrpc":"2.0","id":2,"result":{"isError":false,"content":[{"text":"alice","type":"text"}]}}
+----
+
```

Comment:
This is the "SSE transport" - it would be interesting to see how it works with the streamable HTTP.

Note that your "flow" here, does not do the handshake, so it not very "compliant" - it works because we had to relax that constraint.

Reply:
no streamable http in old version.

Reply:
MCP server does not currently support it as well, Martin had some concerns